azavea/echo-locator
# Add fields to UserProfile

## Overview

This PR adds fields to the `UserProfile` model and also adds a `Destination` model to store destinations associated with each `UserProfile`. This enforces two rules: each profile must have at least one destination and exactly one primary destination.


### Demo
![Screen Shot 2022-04-21 at 2 05 29 PM](https://user-images.githubusercontent.com/77936689/164526947-787df186-d251-4807-9d94-10ec1935a39b.png)

![Screen Shot 2022-04-21 at 2 17 52 PM](https://user-images.githubusercontent.com/77936689/164527013-ff50a5ab-183d-40d4-bb8e-f82eada6de0b.png)

All references to "username" in both the login and create user forms have been renamed to "email".

![Screen Shot 2022-04-21 at 2 17 06 PM](https://user-images.githubusercontent.com/77936689/164526924-31274992-91ea-445a-8a11-5fd5f2c914c9.png)

All fields, including a `TabularInline` with which to add and edit a profile's associated destinations are editable from the admin.

![Screen Shot 2022-04-21 at 2 08 53 PM](https://user-images.githubusercontent.com/77936689/164527386-72b0ce91-6e6b-43cc-89aa-bc8da7d8716b.png)

![Screen Shot 2022-04-21 at 2 09 10 PM](https://user-images.githubusercontent.com/77936689/164527394-6480b50a-3c91-41f0-94b8-89785d99d2f8.png)

Users see an error if they either have no destinations or not exactly one primary destination.

### Notes

This PR has renamed the "username" field to "email" in and removed the actual email field from the admin UI so that users know what to enter and so that the signup experience from the frontend and backend will be the same, e.g. counselors won't have usernames while house seekers don't, etc. This does not change the sign up experience from the command line, where users are still prompted for both an email and username. 

As a note for future work connecting this to the frontend, I've set up a `full_name` field on the `UserProfile` model; however, this is different from the default `User` name fields, which has them as two separate fields. While we can squish first and last names together for storage, separating a full name into separate names can be tricky; this may be worth noting when connecting the form to user and profile creation.


## Testing Instructions

 * Run `./scripts/manage migrate` to update your database.
 * Run `./scripts/server` and navigate to **localhost:8085/admin/**. 
 * Start by creating a new user. Verify that if you enter an invalid email address in the email/username field, it will respond with an error.
 * Then go the User Profile model and add a new profile. 
 * First, test that it is properly enforcing the profile has at least one destination. Complete the form but don't add an address. Click "save" and verify that it responds with an error.
 * Next, text that it is properly enforcing that you have exactly one primary destination. Add an address but do not select that it is a primary destination. Click "save" and verify that it responds with an error.
 * Add another address and select that both are primary destinations. Click "save" and verify that it responds with an error.
 * Finally, remove one of the primary destination checks and click "save". Verify that it saves at it should.


Resolves #248
